x509util/crlcheck/crlcheck.go

```diff
@@ -19,67 +19,203 @@ package main
+	}
+	if err != nil && *strict {
+		return nil, fmt.Errorf("strict CRL parse error: %v", err)
+	}
+	glog.Infof("processing CRL:\n%s", x509util.CRLToString(certList))
+
+	verified := false
+	if len(caCerts) == 0 {
+		glog.Warningf("skipping signature validation as no CA certs available")
+		verified = true
+	}
+	var verifyErr error
+	for _, caCert := range caCerts {
+		if err := caCert.CheckCertificateListSignature(certList); err != nil {
+			verifyErr = err
+		} else {
+			glog.Infof("CRL signature verified against CA cert %q", x509util.NameToString(caCert.Subject))
+			verified = true
+			break
+		}
+	}
+	if !verified {
+		return nil, fmt.Errorf("verification error: %v", verifyErr)
+	}
+	return certList, nil
+}
+
+func processCertArg(filename string, caCerts []*x509.Certificate) error {
+	dataList, err := x509util.ReadPossiblePEMFile(filename, "CERTIFICATE")
+	if err != nil {
+		return err
+	}
+	if len(dataList) == 0 {
+		return fmt.Errorf("no certs found in %s", filename)
+	}
+
+	if len(caCerts) == 0 {
+		// No user-provided CA certs, so use any later entries in the file as possible issuers.
+		for i := 1; i < len(dataList); i++ {
+			issuer, err := x509.ParseCertificate(dataList[i])
 			if err != nil {
-				fmt.Fprintf(os.Stderr, "%s: %v\n", filename, err)
-				errcount++
+				glog.Warningf("failed to parse [%d] in chain: %v", i, err)
 				continue
 			}
-			crls = append(crls, certList)
-			if *verbose {
-				fmt.Print(x509util.CRLToString(certList))
-			}
+			glog.Infof("treating cert [%d] with subject %q as potential issuer", i, x509util.NameToString(issuer.Subject))
+			caCerts = append(caCerts, issuer)
 		}
-		for _, caCert := range caCerts {
-			for _, certList := range crls {
-				if err := caCert.CheckCertificateListSignature(certList); err != nil {
-					fmt.Fprintf(os.Stderr, "%s: Verification error: %v\n", filename, err)
-					errcount++
+	}
+	return processCert(dataList[0], caCerts)
+}
+
+func processCert(data []byte, caCerts []*x509.Certificate) error {
+	client := &http.Client{}
+
+	cert, err := x509.ParseCertificate(data)
+	if err != nil {
+		return fmt.Errorf("certificate parse error: %v", err)
+	}
+	issuer, err := getIssuer(cert, client)
+	if err != nil {
+		glog.Warningf("failed to retrieve issuer for cert: %v", err)
+	}
+	if issuer != nil {
+		glog.Infof("using issuer %q", x509util.NameToString(issuer.Subject))
+		caCerts = append(caCerts, issuer)
+	}
+	expired := false
+	if time.Now().After(cert.NotAfter) {
+		glog.Errorf("certificate is expired (since %v)", cert.NotAfter)
+		expired = true
+	}
+	for _, crldp := range cert.CRLDistributionPoints {
+		glog.Infof("retrieving CRL from %q", crldp)
+		rsp, err := client.Get(crldp)
+		if err != nil || rsp.StatusCode != http.StatusOK {
+			return fmt.Errorf("failed to get CRL from %q: %v", crldp, err)
+		}
+		body, err := ioutil.ReadAll(rsp.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read CRL from %q: %v", crldp, err)
+		}
+		rsp.Body.Close()
```

Comment:
would `defer rsp.Body.Close()` up on L174 be better?

Reply:
It's in a loop so I think it's better to explicit `Close` here.

Reply:
oke doke